src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Helpers.cs

```diff
@@ -0,0 +1,138 @@
+        {
+            Array.Resize(ref a, size);
+#if DEBUG
+            a = a.CloneByteArray();
+#endif
+            return a;
+        }
+
+        public static CmsRecipientCollection DeepCopy(this CmsRecipientCollection recipients)
+        {
+            CmsRecipientCollection recipientsCopy = new CmsRecipientCollection();
+            foreach (CmsRecipient recipient in recipients)
+            {
+                X509Certificate2 originalCert = recipient.Certificate;
+                X509Certificate2 certCopy = new X509Certificate2(originalCert.Handle);
+                CmsRecipient recipientCopy = new CmsRecipient(recipient.RecipientIdentifierType, certCopy);
+                recipientsCopy.Add(recipientCopy);
+                GC.KeepAlive(originalCert);
+            }
+            return recipientsCopy;
+        }
+
+        public static X509Certificate2Collection GetStoreCertificates(StoreName storeName, StoreLocation storeLocation, bool openExistingOnly)
+        {
+            using (X509Store store = new X509Store())
+            {
+                OpenFlags flags = OpenFlags.ReadOnly | OpenFlags.IncludeArchived;
+                if (openExistingOnly)
+                    flags |= OpenFlags.OpenExistingOnly;
+
+                store.Open(flags);
+                X509Certificate2Collection certificates = store.Certificates;
+                return certificates;
+            }
+        }
+
+        public static X509Certificate2 TryFindMatchingCertificate(this X509Certificate2Collection certs, SubjectIdentifier recipientIdentifier)
+        {
+            SubjectIdentifierType recipientIdentifierType = recipientIdentifier.Type;
+            X509Certificate2Collection matches;
+            switch (recipientIdentifierType)
+            {
+                case SubjectIdentifierType.IssuerAndSerialNumber:
+                    {
+                        X509IssuerSerial issuerSerial = (X509IssuerSerial)(recipientIdentifier.Value);
+                        matches = certs.Find(X509FindType.FindByIssuerDistinguishedName, issuerSerial.IssuerName, validOnly: false);
+                        matches = matches.Find(X509FindType.FindBySerialNumber, issuerSerial.SerialNumber, validOnly: false);
+                        break;
+                    }
+
+                case SubjectIdentifierType.SubjectKeyIdentifier:
+                    {
+                        string ski = (string)(recipientIdentifier.Value);
+                        matches = certs.Find(X509FindType.FindBySubjectKeyIdentifier, ski, validOnly: false);
+                        break;
+                    }
+
+                default:
+                    // RecipientInfo's can only be created by this package so if this an invalid type, it's the package's fault.
+                    Debug.Assert(false, $"Invalid recipientIdentifier type: {recipientIdentifierType}");
+                    throw new CryptographicException();
```

Comment:
Compat: Desktop seems to return null in this case.
Coding: Debug.Fail instead of Assert(false)?